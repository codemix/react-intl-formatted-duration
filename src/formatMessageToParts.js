// @flow
import type { IntlShape, MessageDescriptor } from "react-intl";

/**
 * KEY_LITERAL is a symbol used as type for "literal" tokens.
 * It is a Symbol because "type" contains the name of the variable and could collide if it is a string.
 */
export const KEY_LITERAL =
  typeof Symbol === "function" ? Symbol("literal") : "@@literal@@";

export type MessagePart = {
  type: string | typeof KEY_LITERAL,
  value: string
};

/**
 * Like intl.formatMessage but returns an array of token split at the variables.
 * Similar to how <FormattedMessage> handles components in strings
 * https://github.com/yahoo/react-intl/blob/master/src/components/message.js#L58
 *
 * @returns Array
 */
export function formatMessageToParts(
  intl: IntlShape,
  message: MessageDescriptor,
  values?: Object
): MessagePart[] {
  // Note: Since both react-intl and we need a method like this, and has many many other use cases
  // it'd be awesome if react-intl provided an intl.formatMessageToParts() natively instead of having this hack. (through https://github.com/yahoo/intl-messageformat)

  const keys = values != null ? Object.keys(values) : [];

  if (values == null || keys.length === 0) {
    return [
      {
        type: "literal",
        value: intl.formatMessage(message, values)
      }
    ];
  }

  // Creates a token with a random UID that should not be guessable or
  // conflict with other parts of the `message` string.
  const uid = Math.floor(Math.random() * 0x10000000000).toString(16);

  // Replace values variables with our tokens so we know where to inject the variables afterwards
  const tokenizedValues = keys.reduce((tokenized, name) => {
    tokenized[name] = `@__TOKEN-${uid}-${name}__@`;
    return tokenized;
  }, {});

  const formattedMessage = intl.formatMessage(message, tokenizedValues);

  // Split the message into parts so the React Element values captured
  // above can be inserted back into the rendered message.
  const nodes = [];
  const tokenMatcher = new RegExp(`@__TOKEN-${uid}-(.+?)__@`);

  let str = formattedMessage;
  while (str) {
    // find first token
    const match = str.match(tokenMatcher);
    if (!match) {
      // no token left, add everything else as literal
      nodes.push({
        type: KEY_LITERAL,
        value: str
      });

      break;
    }

    const matchStart: number = (match: $FlowFixMe).index;
    const matchLength = match[0].length;
    if (matchStart !== 0) {
      // add all chars before the matched token as a literal node
      const strStart = str.substr(0, matchStart);

      nodes.push({
        type: KEY_LITERAL,
        value: strStart
      });
    }

    // add the matched token as variable node
    const valueKey = match[1];
    nodes.push({
      type: valueKey,
      value: values[valueKey]
    });

    // remove all chars that have been added to `nodes` during this iteration
    str = str.substr(matchStart + matchLength);
  }

  return nodes;
}
