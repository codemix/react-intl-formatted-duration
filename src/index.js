// @flow

/**
 * Unfortunately ReactIntl doesn't support duration yet because it's not implemented
 * in the ECMA specs: https://github.com/yahoo/react-intl/issues/77
 *
 * <FormattedDuration seconds={30} format={EXTENDED_FORMAT} /> becomes <span>30 seconds</span>
 * <FormattedDuration seconds={60} format={EXTENDED_FORMAT} /> becomes <span>1 minute</span>
 * <FormattedDuration seconds={150} format={EXTENDED_FORMAT} /> becomes <span>2 minutes 30 seconds</span>
 * `minutes` and `seconds` are translated
 *
 * <FormattedDuration seconds={30} format={TIMER_FORMAT} /> becomes <span>0:30</span>
 * <FormattedDuration seconds={60} format={TIMER_FORMAT} /> becomes <span>1:00</span>
 * <FormattedDuration seconds={150} format={TIMER_FORMAT} /> becomes <span>2:30</span>
 */

export { default as FormattedDuration, default } from "./FormattedDuration";
export {
  formatDuration,
  formatDurationToParts,
  EXTENDED_FORMAT,
  TIMER_FORMAT
} from "./formatDuration";
