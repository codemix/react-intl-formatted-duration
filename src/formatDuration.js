// @flow
import type { IntlShape, MessageDescriptor } from "react-intl";
import messages from "./messages";
import { formatMessageToParts, KEY_LITERAL } from "./formatMessageToParts";

export const EXTENDED_FORMAT = "EXTENDED_FORMAT";
export const TIMER_FORMAT = "TIMER_FORMAT";

/**
 * Stringifies a duration in a localized format.
 *
 * @param intl       - React Intl intl instance
 * @param seconds    - Duration in seconds
 * @param format     - EXTENDED_FORMAT or TIMER_FORMAT
 * @returns {string} - The duration as a localized string
 */
export function formatDuration(
  intl: IntlShape,
  seconds: number,
  { format }: * = {}
) {
  return formatDurationToParts(intl, seconds, { format })
    .map(token => token.value)
    .join("")
    .trim();
}

/**
 * Generate a duration as an array of tokens. Similar to {@link Intl.DateTimeFormat#formatToParts}
 *
 * @example formatDurationToParts(intl, 150, { format: 'EXTENDED_FORMAT' })
 * [ { type: 'minutes', value: '2' },
 *   { type: 'literal', value: ' ' },
 *   { type: 'unit', value: 'minutes' },
 *   { type: 'literal', value: ' ' },
 *   { type: 'seconds', value: '30' },
 *   { type: 'literal', value: ' ' },
 *   { type: 'unit', value: 'seconds' } ]
 *
 * @example formatDurationToParts(intl, 62, { format: 'TIMER_FORMAT' })
 * [ { type: 'minutes', value: '1' },
 *   { type: 'literal', value: ':' },
 *   { type: 'seconds', value: '02' } ]
 *
 * @param intl      - React Intl intl instance
 * @param seconds   - Duration in seconds
 * @param format    - EXTENDED_FORMAT or TIMER_FORMAT
 * @returns {Array} - The duration string as an array of { type, value } tokens.
 */
export function formatDurationToParts(
  intl: IntlShape,
  seconds: number,
  { format }: * = {}
) {
  const fullDays = Math.floor(seconds / 86400);
  const fullHours = Math.floor(seconds / 3600);
  const hoursLeft = Math.floor(fullHours % 24);
  const fullMinutes = Math.floor(seconds / 60);
  const minutesLeft = Math.floor(fullMinutes % 60);
  const secondsLeft = seconds % 60;

  let valueFormatter;
  let formattingMessage;
  switch (format) {
    case TIMER_FORMAT:
      valueFormatter = padValue;
      formattingMessage = messages.timerFormatting;
      break;
    case EXTENDED_FORMAT:
      valueFormatter = formatUnit;
      formattingMessage = messages.longFormatting;
      break;
    default:
      valueFormatter = formatUnit;
      formattingMessage = format
        ? {
            id: `react-intl-formatted-duration/custom-format/${format}`,
            defaultMessage: format
          }
        : messages.longFormatting;
      break;
  }

  const message =
    intl.messages[formattingMessage.id] ||
    formattingMessage.defaultMessage ||
    "";
  const hasSeconds = message.indexOf("{seconds}") !== -1;
  const hasHours = message.indexOf("{hours}") !== -1;
  const hasDays = message.indexOf("{days}") !== -1;
  const showFullMinutes = !hasHours && !hasDays;
  const showFullHours = !hasDays;

  let tokenizedMessage = formatMessageToParts(intl, formattingMessage, {
    days: hasDays
      ? valueFormatter({
          intl,
          value: fullDays,
          key: "days",
          maxLengthIfPadded: 1,
          showIfZero: false,
          message: messages.daysUnit
        })
      : undefined,
    hours: hasHours
      ? valueFormatter({
          intl,
          value: showFullHours ? fullHours : hoursLeft,
          key: "hours",
          maxLengthIfPadded: 1,
          showIfZero: false,
          message: messages.hoursUnit
        })
      : undefined,
    minutes: valueFormatter({
      intl,
      value:
        (showFullMinutes ? fullMinutes : minutesLeft) +
        (!hasSeconds && secondsLeft >= 30 ? 1 : 0),
      key: "minutes",
      maxLengthIfPadded: 1,
      showIfZero: !hasSeconds && fullHours === 0,
      message: messages.minutesUnit
    }),
    seconds: hasSeconds
      ? valueFormatter({
          intl,
          value: secondsLeft,
          key: "seconds",
          maxLengthIfPadded: 2,
          showIfZero: false,
          message: messages.secondsUnit
        })
      : undefined
  });

  // formatUnit generates more tokenized messages, we need to flatten them
  if (valueFormatter === formatUnit) {
    const flatTokenizedMessage = [];

    for (let i = 0; i < tokenizedMessage.length; i++) {
      const token = tokenizedMessage[i];

      const subTokens = token.value;
      if (!Array.isArray(subTokens)) {
        flatTokenizedMessage.push(token);
        continue;
      }

      for (let j = 0; j < subTokens.length; j++) {
        const subToken = subTokens[j];

        if (subToken.type === "value") {
          subToken.type = token.type;
        }

        flatTokenizedMessage.push(subToken);
      }
    }

    tokenizedMessage = flatTokenizedMessage;
  }

  // Remove trailing spaces (start)
  while (
    tokenizedMessage.length > 0 &&
    tokenizedMessage[0].value.trim() === ""
  ) {
    tokenizedMessage.shift();
  }

  // Remove trailing spaces (end)
  while (
    tokenizedMessage.length > 0 &&
    tokenizedMessage[tokenizedMessage.length - 1].value.trim() === ""
  ) {
    tokenizedMessage.pop();
  }

  // Concat consecutive literals & reduce spaces to 1
  for (let i = 0; i < tokenizedMessage.length - 1; i++) {
    const token = tokenizedMessage[i];
    if (token.type !== KEY_LITERAL) {
      continue;
    }

    const nextToken = tokenizedMessage[i + 1];
    if (nextToken.type !== KEY_LITERAL) {
      token.type = "literal";

      i++; // skip next token too
      continue;
    }

    token.value = (token.value + nextToken.value).replace(/\s\s+/g, " ");
    tokenizedMessage.splice(i + 1, 1);
    i--; // stay on current token.
  }

  return tokenizedMessage;
}

type FormatterProps = {
  intl: IntlShape,
  value: string | number,
  showIfZero?: ?boolean,
  message: MessageDescriptor,
  maxLengthIfPadded: number
};

function formatUnit({ intl, value, showIfZero, message }: FormatterProps) {
  if (!value && !showIfZero) {
    return [];
  }

  value = String(value);

  const unit = intl.formatMessage(message, { value });

  return formatMessageToParts(intl, messages.duration, { value, unit });
}

function padValue({ value, maxLengthIfPadded }: FormatterProps) {
  return `0${value || "0"}`.substr(-maxLengthIfPadded);
}
