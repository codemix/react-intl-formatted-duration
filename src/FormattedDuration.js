// @flow
import * as React from "react";
import PropTypes from "prop-types";
import {
  type IntlShape,
  intlShape,
  FormattedNumber,
  injectIntl
} from "react-intl";
import { formatDurationToParts } from "./formatDuration";
import { type MessagePart } from "./formatMessageToParts";

type Props = {
  intl: IntlShape,
  format?: string,
  seconds: number,
  textComponent: React.ComponentType<any>,
  valueComponent?: React.ComponentType<any>,
  children?: ({ value: string, tokens: MessagePart[] }) => React.Node
};

const printParts = (tokens: MessagePart[]) =>
  tokens
    .map(token => token.value)
    .join("")
    .trim();

class FormattedDuration extends React.Component<Props> {
  render() {
    const { intl, children } = this.props;

    const {
      seconds,
      textComponent: Text = intl.textComponent,
      valueComponent: Value = Text
    } = this.props;

    const tokens = formatDurationToParts(intl, seconds, this.props);
    if (typeof children === "function") {
      return children({ value: printParts(tokens), tokens });
    }

    return (
      <Text>
        {tokens.map(({ type, value }, i) => {
          if (value.trim().length === 0) {
            return value;
          }

          switch (type) {
            case "unit":
            case "literal":
              return <Text key={i}>{value}</Text>;

            default:
              return <Value key={i}>{value}</Value>;
          }
        })}
      </Text>
    );
  }
}

/*
 * This is a hack, react-intl has an internal method "shouldIntlComponentUpdate".
 * FormattedNumber#shouldComponentUpdate does nothing but delegate to that method.
 * We need that method to check if we should update the component.
 *
 * To remove this hack, we should either:
 * - Ask react-intl to expose shouldIntlComponentUpdate and support it officially
 * - Reimplement it, which would have a big decently big impact on the bundle size (https://github.com/yahoo/react-intl/blob/master/src/utils.js)
 */
(FormattedDuration.prototype: $FlowIgnore).shouldComponentUpdate =
  FormattedNumber.prototype.shouldComponentUpdate;

FormattedDuration.contextTypes = { intl: intlShape };

FormattedDuration.propTypes = {
  intl: intlShape.isRequired,
  format: PropTypes.string,
  seconds: PropTypes.number.isRequired,
  textComponent: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
  valueComponent: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
  children: PropTypes.func
};

export default injectIntl(FormattedDuration);
