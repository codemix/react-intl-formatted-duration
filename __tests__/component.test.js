// @flow

import * as React from "react";
import { default as FormattedDuration, EXTENDED_FORMAT } from "../src";
import { mountWithIntl } from "enzyme-react-intl";

function Text(props) {
  return <x-text {...props} />;
}

function Value(props) {
  return <x-value {...props} />;
}

describe("Component", () => {
  it("generates a stable dom", () => {
    const component = mountWithIntl(
      <FormattedDuration
        seconds={61}
        format={EXTENDED_FORMAT}
        textComponent={Text}
        valueComponent={Value}
      />
    );

    expect(component.html()).toMatchSnapshot();
  });

  it("uses a renderprop when supplied", () => {
    const component = mountWithIntl(
      <FormattedDuration seconds={61} format={EXTENDED_FORMAT}>
        {({ value }) => <span>{value}</span>}
      </FormattedDuration>
    );

    expect(component.html()).toMatchSnapshot();
  });

  it("passes tokens to a renderprop and renders them", () => {
    const component = mountWithIntl(
      <FormattedDuration seconds={61}>
        {({ tokens }) => (
          <div>
            {tokens.map(({ type, value }, i) => {
              if (value.trim().length === 0) {
                return value;
              }

              switch (type) {
                case "unit":
                case "literal":
                  return <strong key={i}>{value}</strong>;

                default:
                  return <span key={i}>{value}</span>;
              }
            })}
          </div>
        )}
      </FormattedDuration>
    );

    expect(component.html()).toMatchSnapshot();
  });
});
