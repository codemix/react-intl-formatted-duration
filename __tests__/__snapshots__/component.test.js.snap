// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component generates a stable dom 1`] = `"<x-text><x-value>1</x-value> <x-text>minute</x-text> <x-value>1</x-value> <x-text>second</x-text></x-text>"`;

exports[`Component passes tokens to a renderprop and renders them 1`] = `"<div><span>1</span> <strong>minute</strong> <span>1</span> <strong>second</strong></div>"`;

exports[`Component uses a renderprop when supplied 1`] = `"<span>1 minute 1 second</span>"`;
